Modules/posixmodule.c

```diff
@@ -10126,6 +10126,8 @@ os_putenv_impl(PyObject *module, PyObject *name, PyObject *value)
         goto error;
     }
 
+    /* Use _wputenv() rather than SetEnvironmentVariableW(): _wputenv()
```

Comment:
The search for "=" (line 10102) should begin at index 0, not index 1. Windows allows variable names that begin with 0, but ucrt does not allow it. For example:

```python
>>> os.putenv('spam=', 'eggs')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: illegal environment variable name

>>> os.putenv('=spam', 'eggs')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 22] Invalid argument
```

Reply:
I noticed this surprising difference and I'm trying to understand what should be the correct behavior...

Reply:
Does it work with `'=C'`?

Reply:
Python initializes `nt.environ` from C `_wenviron`, so it follows the long-standing convention of MSVC to exclude hidden variables that begin with "=" and disallowing setting them. We still set them for per-drive working directories (e.g. "=Z:"), but we set them in the process environment only, with `SetEnvironmentVariableW`. For example, see `win32_wchdir` in this module. The drive variables are supposed to be a hidden implementation detail. CMD's `set` doesn't show them, except due to a parsing bug if we run `set ""` or `set "`.

Reply:
If you want to allow setting them, then Python should initialize `nt.environ` from WINAPI `GetEnvironmentStringsW` instead of `_wenviron`, and if the name begins with "=", call `SetEnvironmentVariableW` instead of `_wputenv`.

Reply:
I rewrote my PR to only check for "=" in os.putenv(). os.unsetenv() doesn't check the variable name: just report _wputenv() error. Example:

```
>>> os.unsetenv("key=")

>>> os.unsetenv("=key") 
OSError: [Errno 22] Invalid argument

>>> os.unsetenv("=c:")  
OSError: [Errno 22] Invalid argument
```

By the way, why does os.putenv() explicitly rejects "key=" variable name?

Reply:
What about `os.unsetenv("key=value")`?

> By the way, why does os.putenv() explicitly rejects "key=" variable name?

Because in "key==value" what is a key and what is a value? `key=="key" and value=="=value"` or `key=="key=" and value=="value"`?

Reply:
Windows is so weird. The official documentation says:

> The name of an environment variable cannot include an equal sign (=).

I tested: I **can** set a variable with name "zlong=" (the name contains "="). It is shown by the ``set`` command for example:

```
>>> import os
>>> os.putenv("zlong=", "1")
>>> os.putenv("zlon=g", "2")
>>> os.system("set")
(...)
zlon=g=2
zlong==1
```

Reply:
The `posix_putenv_dict_setitem` call below is unnecessary with `_wputenv`. It copies the string before adding it to `_wenviron`, i.e. it does not comply with POSIX `putenv`.

Reply:
I checked back to even Python 2.7 w/ msvcr90.dll, and even that old version of the CRT copies the string that's passed to `putenv` (via calls to `_calloc_crt` and `strcpy_s`), which I verified by noting the result from `_calloc_crt` and cross-checking with the address returned by `getenv`. The CRT didn't used to copy the string back in the 1990s, but that's ancient history.

Reply:
> The posix_putenv_dict_setitem call below is unnecessary with _wputenv. It copies the string before adding it to _wenviron, i.e. it does not comply with POSIX putenv.

Fixed in commit 0852c7dd52ac42e7843ddfef44571494e4c86070.